File: test/hw_tests/hwtest_scan_compare.cpp
// Copyright (c) 2020-2021 Pilz GmbH & Co. KG
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

#include <rclcpp/rclcpp.hpp>
#include <gtest/gtest.h>

#include <chrono>
#include <functional>
#include <map>
#include <memory>
#include <stdexcept>
#include <string>
#include <thread>

#include <sensor_msgs/msg/laser_scan.hpp>

#include "psen_scan_v2/dist.h"
#include "psen_scan_v2/laserscan_validator.h"

namespace psen_scan_v2_test
{
using namespace std::chrono_literals;

typedef sensor_msgs::msg::LaserScan ScanType;
typedef std::shared_ptr<ScanType const> ScanConstPtr;

static const char* TESTDIR_ENV_VAR{ "HW_TEST_SCAN_COMPARE_TESTDIR" };
static const std::string TEST_DURATION_PARAM_NAME{ "test_duration" };

template <class SubscriptionT, class Rep, class Period>
::testing::AssertionResult isConnected(const std::shared_ptr<SubscriptionT>& subscription,
                                       const std::chrono::duration<Rep, Period>& timeout)
{
  auto start = std::chrono::system_clock::now();
  while (std::chrono::system_clock::now() - start < timeout)
  {
    if (subscription->get_publisher_count() > 0u)
    {
      return ::testing::AssertionSuccess();
    }
    rclcpp::sleep_for(100ns);
  }
  return ::testing::AssertionFailure() << "Failed to establish connection with publisher on topic"
                                       << subscription->get_topic_name();
}

class ScanComparisonTests : public ::testing::Test
{
public:
  void SetUp() override  // Omit using SetUpTestSuite() for googletest below v1.11.0, see
                         // https://github.com/google/googletest/issues/247
  {
    node_ptr_->declare_parameter(TEST_DURATION_PARAM_NAME);
    rclcpp::Parameter test_duration_param = node_ptr_->get_parameter(TEST_DURATION_PARAM_NAME);
    try
    {
      test_duration_ = test_duration_param.as_int();
    }
    catch (const rclcpp::ParameterTypeException& e)
    {
      FAIL() << e.what();
    }

    const char* path{ std::getenv(TESTDIR_ENV_VAR) };
    if (!path)
    {
      FAIL() << "Environment variable " << TESTDIR_ENV_VAR << " not set!";
    }

    RCLCPP_INFO_STREAM(node_ptr_->get_logger(), "Using test directory " << path);
    RCLCPP_INFO_STREAM(node_ptr_->get_logger(), "Using test duration " << test_duration_);

    try
    {
      bins_expected_ = binsFromRosbag(path);
    }
    catch (const std::runtime_error& e)
    {
      FAIL() << "Bag record in " << path
             << " could not be opened. Make sure the directory exists and the you have sufficient rights to open it.";
    }

    // Subscription callbacks have to be processed asynchronously during the test
    async_spinner_ = std::thread([this]() { rclcpp::spin(node_ptr_); });
  }

  void TearDown() override
  {
    rclcpp::shutdown();
    async_spinner_.join();
  }

protected:
  rclcpp::Node::SharedPtr node_ptr_{ std::make_shared<rclcpp::Node>("test_node") };
  std::thread async_spinner_;
  std::map<int16_t, NormalDist> bins_expected_{};
  int test_duration_{ 0 };
};

TEST_F(ScanComparisonTests, simpleCompare)
{
  size_t window_size = 120;  // Keep this high to avoid undersampling

  LaserScanValidator<ScanType> laser_scan_validator(bins_expected_);
  laser_scan_validator.reset();
  // Usage of std::bind is prevented by https://github.com/ros2/rclcpp/issues/273
  auto scan_subscription = node_ptr_->create_subscription<ScanType>(
      "/laser_1/scan", 1000, [&laser_scan_validator, &window_size](ScanConstPtr scan) {
        return laser_scan_validator.scanCb(scan, window_size, 0);
      });

  ASSERT_TRUE(isConnected(scan_subscription, 5s));

  EXPECT_TRUE(laser_scan_validator.waitForResult(test_duration_));
}

}  // namespace psen_scan_v2_test

int main(int argc, char** argv)
{
  rclcpp::init(argc, argv);

  testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

File: test/hw_tests/hwtest_scan_compare.py
# Copyright (c) 2021 Pilz GmbH & Co. KG
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

import os
import unittest

import launch
import launch_ros
import launch_testing

from ament_index_python.packages import get_package_share_directory
from launch.actions import IncludeLaunchDescription
from launch.substitutions import LaunchConfiguration, PathJoinSubstitution
from launch_xml.launch_description_sources import XMLLaunchDescriptionSource

import pytest


def get_optional_env(name, default_value):
    return os.environ[name] if name in os.environ else default_value


@pytest.mark.launch_test
def generate_test_description():
    bringup_launch_descr = XMLLaunchDescriptionSource(
        PathJoinSubstitution([
            get_package_share_directory('psen_scan_v2'),
            'launch',
            'bringup.launch.xml'
        ])
    )

    sensor_ip = get_optional_env('SENSOR_IP', '192.168.0.10')
    host_ip = get_optional_env('HOST_IP', 'auto')
    bringup_launch_args = {'angle_start': '-1.2', 'angle_end': '1.2',
                           'sensor_ip': sensor_ip, 'host_ip': host_ip, 'host_udp_port_data': '55005'}

    test_node = launch_ros.actions.Node(executable=PathJoinSubstitution([LaunchConfiguration(
        'test_binary_dir'), 'hwtest_scan_compare']), name='scan_compare_gtest', parameters=[{'test_duration': LaunchConfiguration('test_duration')}], output='screen')

    return launch.LaunchDescription([
        launch.actions.DeclareLaunchArgument(
            name='test_binary_dir', description='Binary directory of package containing test executables'),
        launch.actions.DeclareLaunchArgument(
            name='test_duration', description='Duration of test'),
        IncludeLaunchDescription(
            bringup_launch_descr,
            launch_arguments=bringup_launch_args.items()
        ),
        test_node,
        launch_testing.util.KeepAliveProc(),
        launch_testing.actions.ReadyToTest(),
    ]), {'scan_compare_gtest': test_node}


class TestScanCompareGTest(unittest.TestCase):

    def test_gtest_terminates(self, proc_info, scan_compare_gtest, test_args):
        # Found no other way of passing test_duration, this only works if it has no default value
        test_duration = int(test_args['test_duration'])
        proc_info.assertWaitForShutdown(process=scan_compare_gtest, timeout=test_duration + 5)


@launch_testing.post_shutdown_test()
class TestOutcome(unittest.TestCase):

    def test_exit_codes(self, proc_info):
        launch_testing.asserts.assertExitCodes(proc_info)
